### ISSUE TITLE
Records and definitional equality broken

### ISSUE THREAD

**Issue by ajrouvoet** [agda/agda#3347]
I've hit a weird thing involving Agda being able to see a definitional equality that looks trivial to me.
Altering the definition of c-==-top to the written alternative makes 'refl' fit into the hole.
With the active definition of c-==-top agda says: `i (⟶-mono (c-==-bot .m)) != zero `

This is the most minimal example I could produce in an hour. Trying to construct a standalone example
from scratch failed to reproduce the bug.

```agda
{- evaluation configurations -}
module _ where
  record Config s v : Set where
    constructor ⟨_,_⟩
    field
      Γ      : Graph s v
      cset   : List (Constraint s v)

  open Config public

{- evaluation -}
data _⟶_ {s v} : ∀ {s' v'} → Config s v → Config s' v' → Set where

  c-==-top :
    ∀ {w t₁ t₂ cs Γ}(θ : TSub s v w) →
    ⟨ Γ , t₁ == t₂ ∷ cs ⟩ ⟶ ⟨ Γ Γ/ θ , cs cs/ θ ⟩
  
  -- alternative
  -- c-==-top :
  --  ∀ {w t₁ t₂ cs Γ}(θ : TSub s v w) →
  --  ⟨ Γ , t₁ == t₂ ∷ cs ⟩ ⟶ ⟨ Γ , [] ⟩

  c-==-bot :
    ∀ {cs Γ}(t₁ t₂ : Term s v) →
    ⟨ Γ , t₁ == t₂ ∷ cs ⟩ ⟶ ⟨ Γ , [] ⟩

module _ where

  record Mono s s' : Set where
    constructor mono
    field
      i   : ℕ
      prf : s' ≡ i + s

  open Mono

  mono-refl  : ∀ {s} → Mono s s
  mono-refl = mono 0 refl

  ⟶-mono : ∀ {s v s' v'}{κ : Config s v}{κ' : Config s' v'} → κ ⟶ κ' →  Mono s s'
  ⟶-mono (c-==-top _)    = mono-refl
  ⟶-mono (c-==-bot _ _)  = mono-refl

  test' : ∀ {s v w cs Γ}{t₁ t₂ : Term s v}{θ : TSub s v w} → ⟶-mono (c-==-top {t₁ = t₁}{t₂}{cs}{Γ} θ) ≡ mono-refl
  test' = refl

  test'' : ∀ {s v cs Γ}{t₁ t₂ : Term s v} → ⟶-mono (c-==-bot {cs = cs}{Γ} t₁ t₂) ≡ mono-refl
  test'' = {!refl!}
```

I'm compiling against Agda 2.5.4.2

**Comment by andreasabel** [comment]
Thanks for reporting!
However, to proceed, we need a self-contained test case which we can run.

**Comment by ajrouvoet** [comment]
Thanks for checking it out. I am trying to create a minimal example, but haven't figured out yet what triggers it. If you have any ideas that can help me narrow it down or issues that come to mind that might be related, that might help.

**Comment by ajrouvoet** [comment]
The following is a complete example that mirrors the entire setup.
But frustratingly this goes through!

At the moment the real project is not public, but if it helps I can share a copy with you or anyone else who wants to have a go at narrowing it down.

-- edit see below for a better example

**Comment by andreasabel** [comment]
Any volunteers for a bug shrinking sprint?  Please contact @ajrouvoet .

**Comment by andreasabel** [comment]
@ajrouvoet  What you could try: make some hidden arguments (e.g. of -->-mono) visible in the pattern matching and see if there are any unexpected blockers to reduction.

**Comment by ajrouvoet** [comment]
Okay, had some more time to narrow it down and that produced the following standalone example.

Notice that the following changes make the refl proof for `c1` go through:

- getting rid of `c0`
- getting rid of the substitution in the `c0` index
- postulating the substitution on `G` function instead of implementing it

Explicating more arguments in either the pattern match of `-->-mono` or in calls made no difference.

```agda
open import Level renaming (zero to ℓz) hiding (lift; Lift)
open import Relation.Binary hiding (_⇒_)

module Bug3347 {ℓ} (𝕃 : DecSetoid ℓz ℓ) where

open import Function as Fun hiding (id)
open import Relation.Binary.PropositionalEquality hiding ([_])

open import Data.Nat
open import Data.List
open import Data.Nat.Properties
open import Data.Vec as Vec using (Vec)
open import Data.Product
open import Data.Fin as Fin using (Fin; zero; suc)
open import Data.Fin.Substitution

data Term (s v : ℕ) : Set where
  var    : Fin v → Term s v

record G s n : Set where
  constructor graph

  field
    datums : Vec (Term s n) s

module App {ℓ s}{T : ℕ → Set ℓ} (l : Lift T (Term s)) where
  open Lift l hiding (var)

  mutual
    infix 8 _/_
    _/_ : ∀ {v w} → Term s v → Sub T v w → Term s w
    var x      / ρ = lift (Vec.lookup x ρ)

module _ {s} where
  tmSubst : TermSubst (Term s)
  tmSubst = record { var = var; app = App._/_ }

  open TermSubst tmSubst hiding (var; subst) public

infixl 8 _Γ/_
_Γ/_ : ∀ {s w v} → G s v → Sub (Term s) v w → G s w
(graph ds) Γ/ ρ = graph (Vec.map (_/ ρ) ds)

module _ where
  record C s v : Set where
    constructor c
    field
      Γ      : G s v

  open C public

data _⟶_ : ∀ {s s' v v'} → C s v → C s' v' → Set where

  c0 :
    ∀ {s v w} Γ (θ : Sub (Term s) v w) →
    c Γ ⟶ c (Γ Γ/ θ)

  c1 :
    ∀ {v s} (Γ : G s v) →
    c Γ ⟶ c Γ

module _ where

  record Mono s s' : Set where
    constructor mono
    field
      i   : ℕ
      prf : s' ≡ i + s

  open Mono

  mono-refl  : ∀ {s} → Mono s s
  mono-refl = mono 0 refl

  ⟶-mono : ∀ {s v s' v'}(κ : C s v)(κ' : C s' v') → κ ⟶ κ' →  Mono s s'
  ⟶-mono _ _ (c0 Γ θ) = mono-refl
  ⟶-mono _ _ (c1 Γ ) = mono-refl

  test' : ∀ {s v w Γ}{θ : Sub (Term s) v w} →
          ⟶-mono _ _ (c0 Γ θ) ≡ mono-refl
  test' = refl

  test'' : ∀ {s v}{Γ : G s v} → (⟶-mono  _ _ (c1 Γ )) ≡ mono-refl
  test'' = {!refl!}
```

**Comment by fredrikNordvallForsberg** [comment]
Giving `refl` in the hole works for me using the current development version (master branch, version `2.6.0-8d6a32c`).

**Comment by ajrouvoet** [comment]
I'm compiling master now to confirm that it is solved entirely there. Maybe someone with more knowledge of what changed can do an educated bisection to figure out where it was solved.

**Comment by ajrouvoet** [comment]
The master branch indeed does the trick :+1:

**Comment by ajrouvoet** [comment]
@fredrikNordvallForsberg is this not something that needs to be addressed in a potential 2.5.4.3 release at some point?

**Comment by fredrikNordvallForsberg** [comment]
I believe the next non critical-bug-fix-only release will be 2.6.0 anyway, so it's probably not worth the effort to try to find and backport the fix.
